### PR TITLE
add SVG support

### LINKED ISSUE #16: SVG primitives
SVGs are the most sane way of expressing shape-like constructs in a react paradigm — c.f. https://github.com/chengyin/react-primitives-svg

Components: https://github.com/airbnb/react-sketchapp/issues?utf8=%E2%9C%93&q=is%3Aissue%20is%3Aopen%20SVG%20Primitives

### SVG components we need
- [ ] Circle
- [ ] Ellipse
- [ ] G
- [ ] Linear Gradient
- [ ] Radial Gradient
- [ ] Line
- [ ] Path
- [ ] Polygon
- [ ] Polyline
- [ ] Rect
- [ ] Symbol
- [ ] Text
- [ ] Use
- [ ] Defs
- [ ] Stop

### LINKED ISSUE #94: Implement Linear Gradient
Hey there!

Do you have any plans about implementation primitive components for linear-gradient and box-shadow?

### LINKED ISSUE #168: Triangle
_I am..._

| -------------------------------------------------------------------------------------------------
| Wondering how to make a triangle
| -------------------------------------------------------------------------------------------------

**Expected behavior:**
I can make a triangle

**Observed behavior:**
I am not sure how to make a triangle, based on the documentation. I would greatly appreciate any help that you can offer. Thank you so much.

### PR DESCRIPTION
**Pull request by mathieudutour**
Fix #16, fix #168, fix #94

It's pretty rough but works.

I'm using the [MSSVGImporter from Sketch to create the layers](https://github.com/airbnb/react-sketchapp/pull/204/files#diff-612a8e68a84e25aaebafd131faf4a1b2R321). Then I'm [parsing the layers back into JSON](https://github.com/airbnb/react-sketchapp/pull/204/files#diff-612a8e68a84e25aaebafd131faf4a1b2R335). It means that we just have to construct an SVG string from the primitives. To do so, I'm [stopping the `flexTreeToJSON` as soon as we find an svg](https://github.com/airbnb/react-sketchapp/pull/204/files#diff-7900d62d7316a6537fe4b09529fe738dR37), and [constructing manually the SVG string](https://github.com/airbnb/react-sketchapp/pull/204/files#diff-8130a2b88d8cedfc2d878a26987893f2R33).

TODO:
- [x] render svg
- [x] docs
- [x] tests
- [x] cleanup propTypes for svg components

screen shot 2017-11-12 at 18 12 29
screen shot 2017-11-12 at 18 12 48

### CODE REVIEW

File: src/debug.js [hunk 1]
```diff
@@ -1,26 +1,70 @@
 /* @flow */
 
 export const dump = (obj: any) => {
-  log('#####################################################################################');
+  log(
+    '#####################################################################################'
+  );
```
Comment (mathieudutour) [inline src/debug.js]: those changes are comming from running `prettier` on the codebase

**Comment by jongold** [comment]
!!! 🎉  — just saw this, let's get a release out of this + the other stuff in the queue tonight / tomorrow

**Comment by gregbenner** [comment]
Goooo @mathieudutour ! Thanks from Canada 👍

**Comment by LincMitch** [comment]
SVG Support WOW!
Keen to have a play with this. When is it likely to be merged? and would this be the "v1.0.0-beta.6" release?

**Comment by LincMitch** [comment]
So I couldn't wait and tried installing the branch. But it depends on `"react-sketchapp": "^1.0.0",` as appose to ` "react-sketchapp": "^1.0.0-beta.5",`.

Do I assume that 1.0.0 is a local version?

**Comment by mathieudutour** [comment]
Teah if you want to try the example locally, you need to build react-Sketchapp and point to it, it’s not published yet

**Comment by artemshitov** [comment]
I couldn't get `ry` and `rx` to work in radial gradients. React Native SVG has them. Do you have plans to implement these?

**Comment by mathieudutour** [comment]
hum it's probably that sketch doesn't support it so can't really implement it in `react-sketchapp`

**Comment by mathieudutour** [comment]
actually, can you try to remove the line https://github.com/airbnb/react-sketchapp/pull/204/files#diff-fe421198ccd089168d8b646f236b4917R26 ? maybe it's just a matter of defining stuff that I shouldn't

**Comment by artemshitov** [comment]
Removing the line didn't help, unfortunately.

I think it just needs some special case in rendering, because Sketch totally can produce transformed radial gradients. Here is JSON for a radial gradient in Sketch:

```json
{
    "gradient": {
        "_class": "gradient",
        "elipseLength": 0.7029145688798363,
        "from": "{0, 0}",
        "gradientType": 1,
        "shouldSmoothenOpacity": false,
        "stops": [
            {
                "_class": "gradientStop",
                "color": {
                    "_class": "color",
                    "alpha": 1,
                    "blue": 1,
                    "green": 1,
                    "red": 1
                },
                "position": 0
            },
            {
                "_class": "gradientStop",
                "color": {
                    "_class": "color",
                    "alpha": 1,
                    "blue": 0,
                    "green": 0,
                    "red": 0
                },
                "position": 1
            }
        ],
        "to": "{0, 1}"
    }
}
```

The gradient looks like this:

image

I guess I could try to implement that once this PR lands. I'm new to the codebase though and don't know whether it is hard to do.

**Comment by mathieudutour** [comment]
I'm not generating the JSON directly, I'm using the SVG importer from Sketch. Do you have a plugin example I could look into?

**Comment by artemshitov** [comment]
Oh, then it won't work. `rx` and `ry` are not in standard SVG, React Native uses these to avoid transforms, most likely. This is how Sketch _exports_ the same gradient:

```xml
<radialGradient cx="0%" cy="0%" fx="0%" fy="0%" r="100%" gradientTransform="translate(0.000000,0.000000),scale(0.606232,1.000000),rotate(90.000000),scale(1.000000,0.702915),translate(-0.000000,-0.000000)" id="radialGradient-1">
```

**Comment by mathieudutour** [comment]
ah well then we probably shouldn't support it, it wouldn't be cross-platform. You can specify a `transform` tho

**Comment by artemshitov** [comment]
Can I specify `gradientTransform`?

**Comment by mathieudutour** [comment]
I'm passing all the props without touching them so yes

**Comment by artemshitov** [comment]
Then it seems like Sketch SVG import doesn't support it yet. I exported the image above to XML from Sketch and then opened in Sketch and Illustrator. Illustrator did just fine, and Sketch ignored all the transformations.

Thank you for your help!

**Comment by LincMitch** [comment]
Thanks for the tip @mathieudutour . 
I think I have a build of react-Sketchapp, as I have a node_modules at root:
**react-sketchapp-f-svg-support/node_modules**

Not sure how to point to it though. Is it a matter of just changing the line:
`"react-sketchapp": "^1.0.0",` to something like `"react-sketchapp": "[PATH TO SOME FILE]",`

**Comment by LincMitch** [comment]
I can see the "Merge branch 'master' into f/svg-support". So I tried downloading this branch again, in the hope that I can get it running - but no luck, just the red error box.

As I got a thumbs up, for my last post. I assume I was on the right track.
If so, what would the path be in the line "react-sketchapp": "[PATH TO SOME FILE]"?

**Comment by bowen31337** [comment]
I am still running into the following errors:
```
(node:79889) UnhandledPromiseRejectionWarning: Unhandled promise rejection (rejection id: 3): TypeError: execFile is not a function
(node:79889) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```
when i execute the command
```
npm run render
```
I used the basic-svg in the examples folder, package.json
```
{
  "name": "basic-svg",
  "version": "1.0.0",
  "description": "",
  "skpm": {
    "main": "basic-svg.sketchplugin",
    "manifest": "src/manifest.json"
  },
  "scripts": {
    "build": "skpm-build",
    "watch": "skpm-build --watch",
    "render": "skpm-build --watch --run",
    "render:once": "skpm-build --run",
    "postinstall": "npm run build && skpm-link"
  },
  "author": "Jon Gold <jon.gold@airbnb.com>",
  "license": "MIT",
  "devDependencies": {
    "@skpm/builder": "^0.2.0"
  },
  "dependencies": {
    "prop-types": "^15.5.8",
    "react": "^15.4.2",
    "react-sketchapp": "1.0.0-beta.5",
    "react-test-renderer": "^15.4.2"
  }
}
```

my node version is v8.2.1, I created the symlink for the branch '/f/svg-support', so the the 'react-sketchapp' in the node_modules redirects to the branch.  so i don't think i have done something wrong, but still can't get it working, can someone help?

**Comment by LincMitch** [comment]
@bowen31337 could this be skpm issue? found this #138

**Comment by LincMitch** [comment]
@bowen31337 I also found this #218

**Comment by mathieudutour** [comment]
@LincMitch, no need for `"react-sketchapp": "[PATH TO SOME FILE]"`, just change the path in `examples/basic-svg/src/my-command.js` to `../../src`

**Comment by LincMitch** [comment]
You legend @jongold for sketchApp and @mathieudutour for this SVG work. Super exciting, Christmas just came early.

**Comment by alampros** [comment]
Hi gents - thank you so much for this!

I wrote (modified, if I'm being honest) a babel plugin that transforms imported SVGs into the SVG primitives from this PR (eg `<g>` => `<Svg.G>`):

https://github.com/alampros/babel-plugin-react-sketchapp-svg

It's almost working (🙌 ) with one weird exception: **the rendered layer in Sketch seems to be stuck with no height**:

svg-height-issue

Here's the code to generate the above:
```jsx
import React from 'react'
import { View } from 'react-sketchapp'
import TestIcon from './check.svg'

export default () => (
  <View width={40} height={40}>
    <TestIcon style={{ width: 40, height: 40 }} width={40} height={40} />
  </View>
)
```
I've tried passing `width` and `height` as both prop attributes and `style` properties to no avail.

Any idea where I'm going wrong or tips on how to debug this?

**Comment by jongold** [comment]
@alampros wow this is awesome!

On vacation without my laptop so I can't debug properly, but could you try setting the `style` prop in `View`?

`<View style={{ height: 40, width: 40 }} >`

**Comment by alampros** [comment]
Thanks @jongold – go enjoy your vacation!

***** For when you return 😄  ******

The `<View>` is being sized correctly (it's the red square above). It's the layer named `Svg` that has `0` height..but somehow has `width`? Could that mean it's related to flex layout?

I *think* the issue is in my babel plugin – most likely not passing props thru somehow. Babel is really powerful, but *reallly* hard to grasp. I'll keep plugging away at it.

**Comment by alampros** [comment]
w00t! I managed to figure it out. It works now! 🎉

**Comment by LincMitch** [comment]
@alampros I'm seeing a similar issue, but not with the babel plugin.  Was it Babel in the end and how did you fix it? 

I'm using **react-primitives-svg** to render the same SVG code to both web and sketch:
`switch (Platform.OS) {
  case "sketch":
    SvgLib = require("react-primitives-svg/lib/core.sketch");
    break;
  case "web":
    SvgLib = require("react-primitives-svg/lib/core.web");
    break;
  default: // React Native should be able to pick up `core.ios.js` or `core.android.js` according to the platform
    // SvgLib = require("react-primitives-svg");
}`

Now the SVG height and width seem to be interpreted differently for Sketch and web.
`<Svg viewBox="0 0 200 200" width="200" height="200" fill="blue">` is more like 20/20 on web, but OK in Sketch.
but...
`<Svg viewBox="0 0 200 200" fill="pink">` renders 95.61px /.47px in Sketch, but OK on web.

So I can't use the same code.

Strange!

**Comment by alampros** [comment]
@LincMitch that is super strange. I ended up injecting an invisible square `<rect>` into the SVGs I was working with in order to retain the correct aspect ratio and used a hard coded height and width. That makes me thing this might be related to the yoga layout engine.

Perhaps we should open a separate ticket since this PR is now merged?

**Comment by LincMitch** [comment]
A new ticket sounds good. Will start one.

**Comment by lmnelson** [comment]
@LincMitch SVG seems to be working for me after specifying height and width on the SVG element. Thanks for the tip. When trying to render to web I'm getting a reference error that the variable `NSFontWeightRegular` in `lib/utils/findFont.js` is not defined. Did you run into this issue?

**Pull request merged**